msm/src/test/mod.rs

```diff
@@ -2,3 +2,228 @@ pub mod columns;
+        let verifies = verify::<_, OpeningProof, BaseSponge, ScalarSponge, N>(
+            domain,
+            &srs,
+            &constraints,
+            &proof,
+        );
+        assert!(!verifies)
+    }
+
+    // Test degree a degree one constraint: X_{0} - X_{1}
+    #[test]
+    fn test_completeness_degree_one() {
+        let mut rng = o1_utils::tests::make_test_rng();
+        const N: usize = 2;
+        let domain_size = 1 << 8;
+
+        let constraints = {
+            let x0 = expr::curr_cell::<Fp>(Column::X(0));
+            let x1 = expr::curr_cell::<Fp>(Column::X(1));
+            vec![x0.clone() - x1]
+        };
+
+        let random_x0s: Vec<Fp> = (0..domain_size).map(|_| Fp::rand(&mut rng)).collect();
+        let exp_x1 = random_x0s.clone();
+        let witness: Witness<N, Vec<Fp>> = Witness {
+            cols: [random_x0s, exp_x1],
+        };
+
+        test_completeness_generic::<N, _>(
+            constraints.clone(),
+            witness.clone(),
+            domain_size,
+            &mut rng,
+        );
+        // FIXME: we would want to allow the prover to make a proof, but the verification must fail.
+        // TODO: Refactorize code in prover to handle a degug or add an adversarial prover.
+        // test_soundness_generic(constraints, witness, domain_size, &mut rng);
+    }
+
+    // Test degree a degree one constraint: X_{0} * X_{0} - X_{1} - X_{2}
+    #[test]
+    fn test_completeness_degree_two() {
+        let mut rng = o1_utils::tests::make_test_rng();
+        const N: usize = 3;
+        let domain_size = 1 << 8;
+
+        let constraints = {
+            let x0 = expr::curr_cell::<Fp>(Column::X(0));
+            let x1 = expr::curr_cell::<Fp>(Column::X(1));
+            let x2 = expr::curr_cell::<Fp>(Column::X(2));
+            vec![x0.clone() * x0.clone() - x1.clone() - x2.clone()]
+        };
+
+        let random_x0s: Vec<Fp> = (0..domain_size).map(|_| Fp::rand(&mut rng)).collect();
+        let random_x1s: Vec<Fp> = (0..domain_size).map(|_| Fp::rand(&mut rng)).collect();
+        let exp_x2 = random_x0s
+            .iter()
+            .zip(random_x1s.iter())
+            .map(|(x0, x1)| (*x0) * (*x0) - x1)
+            .collect::<Vec<Fp>>();
+        let witness: Witness<N, Vec<Fp>> = Witness {
+            cols: [random_x0s, random_x1s, exp_x2],
+        };
+
+        test_completeness_generic::<N, _>(
+            constraints.clone(),
+            witness.clone(),
+            domain_size,
+            &mut rng,
+        );
+        // FIXME: we would want to allow the prover to make a proof, but the verification must fail.
+        // TODO: Refactorize code in prover to handle a degug or add an adversarial prover.
+        // test_soundness_generic(constraints, witness, domain_size, &mut rng);
+    }
+
+    // Test degree a degree one constraint:
+    //   X_{0} * X_{0} * X_{0} \
+    // - 42 * X_{1} * X_{2} \
+    // + X_{3}
+    #[test]
+    fn test_completeness_degree_three() {
```

Comment:
I think all these tests for different constant degree (2,3,6,7) have a lot of commonly shared boilerplate and can be probably made much more concise later on

Reply:
Agreed. It is planned when we can use the generic interpreter, see comment.